Fix warning suppression for config.toml vs config compat symlinks

ijackson opened the pull request:
### What does this PR try to resolve?

Background: the cargo config file is being renamed from `.cargo/config` to `.cargo/config.toml`.  There's code in new cargo to look for both files (for compatibility), to issue a warning when onliy the old filename is found, and also to issue a warning if both files are found.  The warning suggests making a symlink if compatibility with old cargo is wanted.

An attempt was made to detect when both the old and new files exists, but one is a symlink to the other, but as reported in #13667, this code is not effective.  (It would work only if the symlink had the precise absolute pathname that cargo has decided to use for the lookup, which would be an unnatural way to make the link.)

Logically, the warning should appear when both files exist *but are different*.  That is the anomalous situation that will generate confusing behaviour.   By "different" we ought to mean "aren't the very same file".

That's what this MR implements, where possible.  On Unix, we use the information from stat(2).  That's not available on other platforms; on those, we arrange to also tolerate a symlink referring to precisely `config.toml` as a relative pathname, which is also fine, since by definition the target is then in the same directrory as `config`.

Fixes #13667.

### How should we test and review this PR?

I have interleaved the new tests with the commits that support them.  In each case, a functional commit is followed by a test which fails just beforehand.

(This can be observed by experimentally reordering the branch.)

I have also done ad-hoc testing.

### Additional information

I'm making the assumption that a symlink containing a relative path does something sane on Windows.  This assumption may be unwarranted.  If so, "Handle `config` -> `config.toml` (without full path)" needs to be dropped, and the test case needs to be `#[cfg(unix)]`.

But also, in this case, we should probably put some warnings in the stdlib docs!

epage commented:
Please add all test cases in the initial commit, passing.  This should show what the current behavior is.  As the commits are made that change behavior, the test cases will then be updated to pass with the behavior changes.

This has a couple of benefits
- Helps test the tests
- Makes it more clear to anyone reading this what the behavior changes are

epage commented on tests/testsuite/config.rs:
```diff
@@ -309,6 +308,13 @@ f1 = 1
+    config_ambiguous_filename_symlink_doesnt_warn_general(|| {
+        symlink_config_to_config_toml();
+    });
+}
```
nit: personally i would rather duplicate the tests than do this.  This makes it harder to follow what the tests are doing and to evolve them over time.

ijackson replied:
Gosh.  OK.

ijackson replied:
Now done.

epage commented on src/cargo/util/context/mod.rs:
```diff
@@ -1546,7 +1546,10 @@ impl GlobalContext {
                 // get a warning.
                 let skip_warning = if let Ok(target_path) = fs::read_link(&possible) {
-                    target_path == possible_with_extension
+                    target_path == possible_with_extension ||
+                        // allow `config` -> `config.toml`, without the parent dir(s),
+                        // which is how you'd naturally do it
+                        target_path == Path::new(&filename_with_extension)
```
is there a reason not to just make both paths absolute and compare them?

ijackson replied:
Yes.  That is considerably slower (which may not matter very much, but this happens every time cargo starts up).

Perhaps that could be done in the  non-Unix case, but surely the perf problem is worse on Windows (and symlinks aren't really a thing there anyway).

epage replied:
And yes, I had overlooked it, `same-file` should instead be used

ijackson replied:
> And yes, I had overlooked it, same-file should instead be used

OK, thanks, willdo.

weihanglo commented on src/cargo/util/context/mod.rs:
```diff
@@ -1544,8 +1545,34 @@ impl GlobalContext {
+                ) {
+                    use std::os::unix::fs::MetadataExt;
+                    let devino = |md: fs::Metadata| (md.dev(), md.ino());
+                    devino(metadata_1) == devino(metadata_2)
```
Cargo depends on [`same-file`](https://crates.io/crates/same-file) transitively via `cargo-util` btw. We might want to make use of it instead.

ijackson replied:
Interesting.  I was all ready to say "oh I should use that instead".

But I RTFM and `samefile` is unreliable on Windows.  I don't really have an opinion about that (other than "it is impossible to write reliable software on Windows, anyway").  Perhaps it's OK for this to be unreliable on Windows since the only misbehaviour would be to suppress a warning that ought to have been issued.

I would be happy to change this to use `samefile` everywhere.  The documentation is clear about the algorithm on Unix, and it's a reliable one.

LMK what you prefer.

ijackson commented:
> Please add all test cases in the initial commit, passing. This should show what the current behavior is. As the commits are made that change behavior, the test cases will then be updated to pass with the behavior changes.

OK.  When should I rebase the branch then?  Normally I wouldn't do that without an OK from the reviewers.

epage commented:
> OK. When should I rebase the branch then? Normally I wouldn't do that without an OK from the reviewers.

yes, please

ijackson commented:
OK.  Thanks everyone for the comments.  LMK about the `samefile` thing, and I'll probably provide a revised version tomorrow.

ijackson commented:
I've revised this and rebased/reordered it, as discussed.

While I was writing the code, I felt dirty writing `if let Ok(...) = `, because that's an ignored error bug.  But I don't think this MR is the right time to address this, so I just added a comment about it.  Ideally we'd change this here, but also  use clippy to forbid use of the buggy `exists` function throughout the codebase.  If folks agree, I could do that as a followup.

ijackson commented:
Err, I have discovered a mistake.  Give me a moment, will rewind this again.  Sorry.

ijackson commented:
I think this is ready now.  CI failures are unrelated.

epage commented on src/cargo/util/context/mod.rs:
```diff
@@ -1537,6 +1537,11 @@ impl GlobalContext {
         let possible = dir.join(filename_without_extension);
         let possible_with_extension = dir.join(format!("{}.toml", filename_without_extension));
 
+        // Arguably this is wrong, because it ignores errors.
+        // I.e. it is "error-prone", as discussed in the docs for `std::path::Path::exists`.
+        //
+        // Instead, if we got an error, we should bail, but perhaps
+        // that might be too much risk of being a breaking change.
```
imo code comments shouldn't be used for a backlog.  Please pull out this commit

epage replied:
If you have a reproduction case for a bug, feel free to open an Issue.

ijackson commented on src/cargo/util/context/mod.rs:
```diff
@@ -1537,36 +1537,32 @@ impl GlobalContext {
-
-                if !skip_warning {
-                    if possible_with_extension.exists() {
+                    if possible_handle != possible_with_extension_handle {
```
I did this deliberately.  I think it's fine for either side to be a symlink.  Let me explain:

The reason we are warning is as follows: if the user has both these files, and they are different, then different versions of cargo might behave differently due to reading different files.  That would be quite confusing.

If both files are the same, whichever way round they are symlinked, then this risk doesn't arise.

So insisting on the symlink being made in one particular direction would just be being pernickety for no good reason.  (And also, our current  warning message, inviting the user to make a symlink, is ambiguous, because the phrase "link A to B" might mean "make A -> B" or "make B -> A".)

If we *do* want to insist on the symlink being made in one particular direction (which I think we don't) then we certainly ought to have a bespoke error message for this situation.

We would also then have to decide what we think if *both* files are symlinks referring, ultimately, to the same file, somewhere else.  Currently we're happy with this, as I think we should be.

epage replied:
A change in behavior like that should be called out explicitly in an easy-to-discover way in the PR description

bors commented:
:pushpin: Commit 2f168383858fac36b5829e777879474af6487eb9 has been approved by `epage`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/cargo) for this repository.

<!-- @bors r=epage 2f168383858fac36b5829e777879474af6487eb9 -->
<!-- homu: {"type":"Approved","sha":"2f168383858fac36b5829e777879474af6487eb9","approver":"epage","queue":"https://bors.rust-lang.org/homu/queue/cargo"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang/cargo/actions/runs/8820410733/job/24214122954)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang/cargo/actions/runs/8820410733/job/24214122954","builder_name":"checks-actions"} -->

bors commented:
:hourglass: Testing commit 2f168383858fac36b5829e777879474af6487eb9 with merge 52dae0c1f650ac8a376df7ee402cb9a3ffd147d6...
<!-- homu: {"type":"BuildStarted","head_sha":"2f168383858fac36b5829e777879474af6487eb9","merge_sha":"52dae0c1f650ac8a376df7ee402cb9a3ffd147d6"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang/cargo/actions/runs/8820856400/job/24216370573)
Approved by: epage
Pushing 52dae0c1f650ac8a376df7ee402cb9a3ffd147d6 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"epage","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang/cargo/actions/runs/8820856400/job/24216370573"},"merge_sha":"52dae0c1f650ac8a376df7ee402cb9a3ffd147d6"} -->
